### PR TITLE
Upgrade `rules_python` to 0.27.0

### PR DESCRIPTION
**Pull request by onelxj**

### CODE REVIEW

**Comment by onelxj** [comment]
@rjhuijsman any thoughts about the file structure we have for bazel?

File: WORKSPACE.bazel [hunk 1]
```diff
@@ -7,3 +7,18 @@ repos(external = False)
...
+# Initialize repositories for all packages in requirements.txt.
+# TODO: Move it to a separate function (file).
+load("@python3_10_12//:defs.bzl", "interpreter")
+load("@rules_python//python:pip.bzl", "pip_parse")
+
+pip_parse(
+    name = "pyprotoc_plugin_deps",
+    python_interpreter_target = interpreter,
+    requirements_lock = "@com_github_reboot_dev_pyprotoc_plugin//:requirements.txt",
+)
```
Comment (rjhuijsman) [inline WORKSPACE.bazel:20]: I think the key to good organization here is to treat this `pip_parse` as a repo-creating instruction. So we should organize it with the "repos", not the "deps".

I'd suggest creating a `python_repos.bzl` file and doing this work in there. Then in this `WORKSPACE.bazel` file calling `python_repos()` either just before or just after `repos()`.

**Review by rjhuijsman — CHANGES_REQUESTED** [review]
Cool, I like the new `requirements.in`->`requirements_lock.txt`! 👍🏻 

The tricky part in creating the order things are loaded is: when do you initialize the toolchain? There's not a great place for it in our traditional repos-deps organization.

The `rules_python` sort of dictates the answer: it has to happen after getting the `rules_python` repo, but before setting up the PyPI repos. And (see my inline comment about what "repos" and "deps" means) that means the toolchain should be in the middle of the "repos" steps.

So here's an order I think could work (not meant to compile, just to give the structure).

In the existing `repos.bzl`:
```
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

def repos():
    ...
	http_archive(
	    name = "rules_python",
	    sha256 = SHA,
	    strip_prefix = "rules_python-{}".format(VERSION),
	    url = "https://github.com/bazelbuild/rules_python/releases/download/{}/rules_python-{}.tar.gz".format(VERSION,VERSION),
	)
```

In a new `py_toolchains.bzl`:
```
load("@rules_python//python:repositories.bzl", "py_repositories", "python_register_toolchains")

def toolchains():
  py_repositories()

  python_register_toolchains(
	    name = "python_3_10_...",
	    python_version = "3.10_...",
	)
```

And with `pypi_repos.bzl` as you already have it.

Then the order can be:
1. `repos.bzl`
2. `py_toolchains.bzl`
3. `pypi_repos.bzl`
4. `deps.bzl` (very small now that all Python stuff has moved elsewhere)

File: bazel/py_toolchains.bzl [hunk 2]
```diff
@@ -0,0 +1,12 @@
+"""Defines a local python toolchain for use in rules_python."""
+
+load("@rules_python//python:repositories.bzl", "py_repositories", "python_register_toolchains")
+
+def toolchains(repo_mapping = {}):
```
Comment (rjhuijsman) [inline bazel/py_toolchains.bzl:5]: Let's make it extra obvious and call this `py_toolchains`. Then also call it that in the `WORKSPACE.bazel`.

File: WORKSPACE.bazel [hunk 3]
```diff
@@ -4,6 +4,14 @@ load("//bazel:repos.bzl", "repos")
 
 repos(external = False)
 
+load("//bazel:py_toolchains.bzl", "toolchains")
+
+toolchains()
+
+load("//bazel:pypi_repos.bzl", pyprotoc_plugin_pypi_repos = "pypi_repos")
```
Comment (rjhuijsman) [inline WORKSPACE.bazel:11]: Everything in this repo is about `pyprotoc_plugin`, so it's not super useful to add that prefix to the name. I'd keep it at just `pypi_repos`.

File: bazel/deps.bzl [hunk 4]
```diff
@@ -1,13 +1,8 @@
...
 
 def deps(repo_mapping = {}):
-    pip_install(
-        name = "pyprotoc_plugin_deps",
-        requirements = "@com_github_reboot_dev_pyprotoc_plugin//:requirements.txt",
-        repo_mapping = repo_mapping,
-    )
-
     protobuf_deps()
+    install_deps(repo_mapping = repo_mapping)
```
Comment (rjhuijsman) [inline bazel/deps.bzl:8]: Here the name `install_deps` makes it a little harder to notice that this is the PyPI deps, so I'd consider importing `install_deps` as `pypi_deps`.

File: bazel/pypi_repos.bzl [hunk 5]
```diff
@@ -0,0 +1,14 @@
+"""Defines a local PyPI repository for the pyprotoc_plugin, the targets
+should be installed later in WORKSPACE file, to be accessible.
+`pip_parse` takes a `requirements_lock` file as input, which is
+generated by `bazel run :requirements.update` command."""
```
Comment (rjhuijsman) [inline bazel/pypi_repos.bzl:4]: Some wording suggestions:

```suggestion
"""Defines local PyPI-based repositories. Their deps
should be installed later in WORKSPACE file to be accessible.
`pip_parse` takes a `requirements_lock` file as input, which is
generated by the `bazel run :requirements.update` command."""
```

**Review by rjhuijsman — APPROVED** [review]
Great, thank you! Just a few naming suggestions - then let's ship it! 👍🏻

**Pull request merged**